contracts/Raffles.cdc

```diff
@@ -0,0 +1,153 @@
+import "MetadataViews"
+
+pub contract Raffles {
+    pub let ManagerStoragePath: StoragePath
+    pub let ManagerPublicPath: PublicPath
+
+    pub event RaffleCreated(address: Address, raffleID: UInt64, sourceType: Type)
+    pub event RaffleDrawn(address: Address, raffleID: UInt64, sourceType: Type, index: Int, value: String)
+
+    pub struct Details {
+        pub let start: UInt64?
+        pub let end: UInt64?
+        pub let display: MetadataViews.Display
+
+        init(
+            _ start: UInt64?,
+            _ end: UInt64,
+            _ display: MetadataViews.Display
+            
+        ) {
+            self.start = start
+            self.end = end
+            self.display = display
+        }
+    }
+
+    pub resource interface RafflePublic {
+        pub fun getEntryAt(index: Int): AnyStruct
+    }
+
+    pub resource Raffle: RafflePublic, MetadataViews.Resolver {
+        pub let source: @{RaffleSource}
+        pub let details: Details
+
+        pub fun draw(): Int {
+            return self.source.draw()
+        }
+
+        pub fun getEntryAt(index: Int): AnyStruct {
+            return self.source.getEntryAt(index: index)
+        }
+
+        pub fun getViews(): [Type] {
+            return [
+                Type<MetadataViews.Display>()
+            ]
+        }
+
+        pub fun resolveView(_ type: Type): AnyStruct? {
+            switch type {
+                case Type<MetadataViews.Display>():
+                    return self.details.display
+            }
+
+            return nil
+        }
+
+        init(
+            source: @{RaffleSource},
+            details: Details
+        ) {
+            self.source <- source
+            self.details = details
+        }
+
+        destroy() {
+            destroy self.source
+        }
+    }
+
+    pub resource interface RaffleSource {
+        pub fun draw(): Int
+        pub fun getEntryAt(index: Int): AnyStruct
+    }
+
+    pub resource AddressRaffleSource: RaffleSource {
+        pub let addresses: [Address]
+
+        pub fun draw(): Int {
+            let r = revertibleRandom()
+            return Int(UInt64(self.addresses.length) % r)
```

Comment:
Here should be `return Int(r % UInt64(self.addresses.length))` right?

Reply:
ahhh good catch! Yes, these need to be flipped